### PR TITLE
Signed operations as syscalls

### PR DESCRIPTION
**Pull request by unzvfu**
This PR implements `SDIV`, `SMOD`, `SIGNEXTEND`, `BYTE`, `SAR`, `SLT` and `SGT` as kernel syscalls. Also links `EXP` to the syscall subsystem.

Hat tip to @nbgl for several of the syscall implementations.

### CODE REVIEW

File: evm/src/cpu/kernel/tests/signed_syscalls.rs [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+/// Given a U256 x, interpret it as a signed 256-bit number and return
+/// the pair abs(x) and sign(x), where sign(x) = 1 if x < 0, and 0
+/// otherwise. NB: abs(x) is interpreted as an unsigned value, so
+/// u256_abs_sgn(-2^255) = (2^255, -1).
+fn u256_abs_sgn(x: U256) -> (U256, bool) {
+    let is_neg = x.bit(255);
+
+    // negate x if it's negative
+    let x = if is_neg {
+        U256::zero().overflowing_sub(x).0
```
Comment (nbgl) [inline evm/src/cpu/kernel/tests/signed_syscalls.rs:58]: Nit:
```suggestion
        x.overflowing_neg().0
```

**Review by nbgl — COMMENTED** [review]
Looks good!! A few comments, but nothing very profound :)

File: evm/src/cpu/kernel/asm/exp.asm [hunk 2]
```diff
@@ -73,4 +73,43 @@ recursion_return:
...
+    // stack: shift_bits, x, e, return_info
+    %div_const(8)
+    // stack: index_of_nz_byte := shift_bits / 8, x, e, return_info
+    dup3
+    iszero
+    push 1
+    sub
+    // stack: e_non_zero := 1 - e_is_zero, index_of_nz_byte, x, e, return_info
+    add
+    // stack: byte_size_of_e := index_of_nz_byte + e_non_zero, x, e, return_info
```
[the remark this answers is not in the document]
Comment (unzvfu) [reply evm/src/cpu/kernel/asm/exp.asm:100]: Oh, that's much neater; thanks!

File: evm/src/cpu/kernel/tests/signed_syscalls.rs [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+use crate::cpu::kernel::aggregator::KERNEL;
+use crate::cpu::kernel::interpreter::Interpreter;
+
+/// Generate a list of inputs suitable for testing the signed operations
+///
+/// The result includes 0, ±1, ±2^(16i ± 1) for i = 0..15, and ±2^255
+/// and then each of those ±1. No attempt has been made to avoid
+/// duplicates. Total length is 288.
+fn test_inputs() -> Vec<U256> {
+    let mut res = vec![U256::zero(), U256::one()];
```
[the remark this answers is not in the document]
Comment (unzvfu) [reply evm/src/cpu/kernel/tests/signed_syscalls.rs:12]: Yeah, I left that in to make it explicit, but probably unnecessary.

File: evm/src/cpu/kernel/tests/signed_syscalls.rs [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    for i in 1..16 {
+        res.push(U256::one() << (16 * i));
+        res.push(U256::one() << (16 * i + 1));
+        res.push(U256::one() << (16 * i - 1));
+    }
+    res.push(U256::one() << 255);
+
+    let n = res.len();
+    for i in 0..n {
```
[the remark this answers is not in the document]
Comment (unzvfu) [reply evm/src/cpu/kernel/tests/signed_syscalls.rs:21]: As above, I left that in to make it explicit, but agree it's unnecessary.

File: evm/src/cpu/kernel/asm/exp.asm [hunk 5]
```diff
@@ -73,4 +73,43 @@ recursion_return:
...
+    %mul_const(@GAS_EXPBYTE)
+    %add_const(@GAS_EXP)
+    // stack: 10 + 50 * byte_size_of_e, x, e, return_info
+    %charge_gas
```
[the remark this answers is not in the document]
Comment (unzvfu) [reply evm/src/cpu/kernel/asm/exp.asm:104]: Oops, good catch!

**Review by nbgl — APPROVED** [review]
looks good!!

**Pull request merged**